File: system-http-client/src/test/java/com/levnovikov/core_network/api_provider/ApiProviderTest.java
package com.levnovikov.core_network.api_provider;

import com.google.gson.Gson;
import com.levnovikov.core_network.HttpClient;
import com.levnovikov.core_network.request.Request;
import com.levnovikov.core_network.response.Response;
import com.levnovikov.core_network.response.ResponseBody;

import junit.framework.Assert;

import org.junit.Test;

import java.io.ByteArrayInputStream;
import java.nio.charset.StandardCharsets;

import io.reactivex.Single;
import io.reactivex.observers.TestObserver;

/**
 * Author: lev.novikov
 * Date: 27/3/18.
 */
public class ApiProviderTest {

    class UserDataRequest {
        private String id;

        public String getId() {
            return id;
        }

        public void setId(String id) {
            this.id = id;
        }
    }
    class UserDataResponse {
        private String name;
        private int age;

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        public int getAge() {
            return age;
        }

        public void setAge(int age) {
            this.age = age;
        }
    }

    @Test
    public void makeRequest() throws Exception {
        EntityConverter converter = new EntityConverter() {

            private Gson gson = new Gson();

            @Override
            public <T> String convertTo(T entity) {
                return gson.toJson(entity);
            }

            @Override
            public <T> T convertFrom(String string, Class<T> responseClass) {
                return gson.fromJson(string, responseClass);
            }
        };

        HttpClient client = request -> {
            UserDataResponse response = new UserDataResponse();
            response.age = 20;
            response.name = "User name";
            ResponseBody responseBody = new ResponseBody(
                    new ByteArrayInputStream(converter.convertTo(response).getBytes(StandardCharsets.UTF_8)), "utf-8", "");
            return Single.just(new Response.Builder().body(responseBody).build());
        };

        ApiProvider provider = new ApiProvider.Builder()
                .baseUrl("https://google.com")
                .client(client)
                .converter(converter)
                .build();

        UserDataRequest request = new UserDataRequest();
        request.id = "5";
        TestObserver<UserDataResponse> observer = provider
                .makeRequest(Request.Method.POST, "/sdfsd/", request, UserDataResponse.class).test();
        observer.assertValueCount(1);
        UserDataResponse response = observer.values().get(0);
        Assert.assertEquals(20, response.age);
        Assert.assertEquals("User name", response.name);
    }

}

File: system-http-client/src/test/java/com/levnovikov/core_network/api_provider/ApiProviderTest.kt
package com.levnovikov.core_network.api_provider

import com.levnovikov.core_network.HttpClient
import com.levnovikov.core_network.request.Request
import com.levnovikov.core_network.response.Response
import com.levnovikov.core_network.response.ResponseBody
import junit.framework.Assert
import org.junit.Test
import java.io.ByteArrayInputStream
import java.nio.charset.StandardCharsets
import kotlin.reflect.KClass

/**
 * Author: lev.novikov
 * Date: 27/3/18.
 */
class ApiProviderTest {

    internal data class UserDataRequest(var id: String? = null)

    internal data class UserDataResponse(var name: String? = null, var age: Int = 0)

    @Test
    @Throws(Exception::class)
    fun makeRequest() {
        val converter = object : EntityConverter {

            override fun <T : Any> convertTo(entity: T): String {
                return when (entity) {
                    is ApiProviderTest.UserDataRequest -> "{'id': ${entity.id}}"
                    is UserDataResponse -> "{'name':'User name', 'age': 20}"
                    else -> throw UnsupportedOperationException("Not supported")
                }
            }

            @Suppress("UNCHECKED_CAST")
            override fun <T : Any> convertFrom(string: String, responseClass: KClass<T>): T {
                return when (responseClass) {
                    UserDataResponse::class -> UserDataResponse("User name", 20) as T
                    else -> throw UnsupportedOperationException("Not supported")
                }
            }
        }

        val client: HttpClient = object : HttpClient {
            override fun makeCall(request: Request): Response {
                val response = UserDataResponse()
                response.age = 20
                response.name = "User name"
                val responseBody = ResponseBody(
                        ByteArrayInputStream(converter.convertTo(response).toByteArray(StandardCharsets.UTF_8)), "utf-8", "")
                return Response.Builder().body(responseBody).build()
            }
        }

        val provider = ApiProvider.Builder()
                .baseUrl("https://google.com")
                .client(client)
                .contentType("text/html")
                .converter(converter)
                .build()

        val request = UserDataRequest()
        request.id = "5"
        val response = provider
                .makeRequest(Request.Method.POST, "/sdfsd/", request, null, UserDataResponse::class)
        Assert.assertEquals(20, response.age)
        Assert.assertEquals("User name", response.name)
    }

}